File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/javascript/stack.js
class Stack {

  constructor() {
    this.stack = [];
    // this is an arbitrary value to make testing easier
    // 1,024 would be way too high to test!
    this.limit = 10;
    this.counter = 0; 
  }

  // add item to top of stack if not full
  // if full throw error
  push(item) {
    if (this.isFull()) { 
   console.log(error)
    }
    // else stack.push(item)
    // counter++ 
    else {this.stack[this.stack.length] = item}
  }

  // remove item from top of stack and return it
  pop() {
    let item = this.stack.splice(this.stack.length - 1)
    return item[0]
  }

  // return item at top of stack without removing it
  peek() {
    return this.stack[this.stack.length - 1]
  }

  // return true if stack is empty, otherwise false
  isEmpty() {
    if (this.stack.length == 0) { 
    return true } else { return false }
  }

  // return true if stack is full, otherwise false
  isFull() {
    if (this.stack.length == this.limit) {
      return true } else {return false}
      
    }
  

  // return number of items in stack
  size() {
    return this.stack.length
  }

  // return -1 if item not in stack, otherwise integer representing 
  // how far it is from the top
  search(target) {
  // let found = this.stack.find(target) 
  // if (found.isInteger()) {return this.stack.size - 1 - found }  else { return - 1} 
  // }
    for (let i = 0; i < this.stack.length; i++){
      if (this.stack[i] == target) { return this.stack.length - 1 - i } 
    }
    return -1 
    } 

  // print contents of stack: do not return the stack itself!
  print() {
    for (let i = 0; i < this.stack.length; i++){
      console.log(this.stack[i])
    }
  }
}

if (require.main === module) {
  // add your own tests in here
}

module.exports = Stack;

File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/ruby/stack.rb
class Stack
  attr_reader :limit

  def initialize
    @stack = []
    # this is an arbitrary value to make testing easier
    # 1,024 would be way too high to test!
    @limit = 10
  end

  # add item to top of stack if not full
  # if full, throw error
  def push(item)
 
   if isFull? 
    raise 'Stack is full' 
   else
    @stack.push(item)
   end
  end

  # remove item from top of stack and return it
  def pop
    return @stack.pop #.pop in ruby removes and returns last element 
  end

  # return item at top of stack without removing it
  def peek
     @stack.last
  end

  # return true if stack is empty, otherwise false
  def isEmpty?
    @stack.size.zero?
  end

  # return true if stack is full, otherwise false
  def isFull?
    @stack.size == @limit
  end

  # return number of items in stack
  def size
   @stack.length #.length & .size = same 
  end

  # return -1 if item not in stack, otherwise integer representing 
  # how far it is from the top
  def search(target)
     @stack.each_with_index do |item, index| #each with index iterates & gets element as well as index 
     return size - index - 1 if item == target 
     end 
      -1 
  end 


  # print contents of stack: do not return the stack itself!
  def print
    puts @stack.join(' , ')
  end
end

if __FILE__ == $PROGRAM_NAME
  # Don't forget to add your tests!
end

